router/src/dml_handlers/retention_validator.rs

```diff
@@ -0,0 +1,131 @@
+}
+
+impl<C> RetentionValidator<C> {
+    /// Initialise a new [`RetentionValidator`], rejecting time outside retention period
+    pub fn new(catalog: Arc<dyn Catalog>, cache: C) -> Self {
+        Self {
+            catalog,
+            cache,
+            time_provider: Default::default(),
+        }
+    }
+}
+
+#[async_trait]
+impl<C> DmlHandler for RetentionValidator<C>
+where
+    C: NamespaceCache,
+{
+    type WriteError = RetentionError;
+    type DeleteError = RetentionError;
+
+    type WriteInput = HashMap<TableId, (String, MutableBatch)>;
+    type WriteOutput = Self::WriteInput;
+
+    /// Partition the per-table [`MutableBatch`].
+    async fn write(
+        &self,
+        namespace: &NamespaceName<'static>,
+        namespace_id: NamespaceId,
+        batch: Self::WriteInput,
+        _span_ctx: Option<SpanContext>,
+    ) -> Result<Self::WriteOutput, Self::WriteError> {
+        let mut repos = self.catalog.repositories().await;
+
+        // Load the namespace schema from the cache, falling back to pulling it
+        // from the global catalog (if it exists).
+        let schema = self.cache.get_schema(namespace);
+        let schema = match schema {
+            Some(v) => v,
+            None => {
+                // Pull the schema from the global catalog or error if it does
+                // not exist.
+                let schema = get_schema_by_name(namespace, repos.deref_mut())
+                    .await
+                    .map_err(|e| {
+                        warn!(
+                            error=%e,
+                            %namespace,
+                            %namespace_id,
+                            "failed to retrieve namespace schema"
+                        );
+                        RetentionError::NamespaceLookup(e)
+                    })
+                    .map(Arc::new)?;
+
+                self.cache
+                    .put_schema(namespace.clone(), Arc::clone(&schema));
+
+                trace!(%namespace, "schema cache populated");
+                schema
+            }
+        };
+
+        // retention is not infinte, validate all lines of a write are within the retention period
+        // schema.retention_period_ns.map(|retention_period_ns| {
+        if let Some(retention_period_ns) = schema.retention_period_ns {
+            let min_retention = self.time_provider.now().timestamp_nanos() - retention_period_ns;
+            // batch is a HashMap<TableId, (String, MutableBatch)>
+            for (_table_id, (_table_name, batch)) in &batch {
+                let time_summary = batch.timestamp_summary();
+                if let Some(time_summary) = time_summary {
+                    let min = time_summary.stats.min;
+                    if let Some(min) = min {
+                        if min < min_retention {
+                            return Err(RetentionError::OutsideRetention);
+                        }
+                    }
+                }
```

Comment:
We can simplify this block for the reader:

```suggestion
                                if let Some(min) = batch.timestamp_summary().and_then(|v| v.stats.min) {
                    if min < min_retention {
                        return Err(RetentionError::OutsideRetention);
                    }
                }
```

That removes some redundant nesting and value matching which makes it a little easier to read :+1: